### PR TITLE
Add support for mm:ss.ms time format

### PR DESCRIPTION
**Pull request by Dananji**
Changes in this PR:
1. Support for hh:mm:ss(.ss), mm:ss(.ss), and ss(.ss) time formats
2. Removed moment.js dependency
3. Quick fix for timespan validation when dragging handles in waveform

### CODE REVIEW

File: dist/services/__test__/StructuralMetadataUtils.test.js [hunk 1]
```diff
@@ -132,8 +124,24 @@ describe('StructuralMetadataUtils class', () => {
...
-    test('when end time exceeds (00:38:58.000) file duration (00:28:58.950)', () => {
+    test('when time is in mm:ss.ms (00:41.45) format', () => {
+      const timespan = smu.findItem('123a-456b-789c-1d', structure);
+      expect(timespan.begin).toEqual('00:00:41.450');
+    });
+    test('when time is in mm:ss.ms (10:41.45) format', () => {
+      const timespan = smu.findItem('123a-456b-789c-1d', structure);
+      expect(timespan.end).toEqual('00:10:41.450');
+    });
+    test('when time is in mm.ss (15.30) format', () => {
```
Comment (cjcolvar) [inline dist/services/__test__/StructuralMetadataUtils.test.js]: I'm not sure if this is a valid value that needs to be supported.  Have you seen this in an old structure?

**Comment by bkeese** [comment]
There is a much simpler algorithm that we used to use for this (though it wouldn't support mm.ss, which is questionable anyway):
```function timeCodeToSeconds(hh_mm_ss){
    var tc_array = hh_mm_ss.split(":").reverse(),
      tc_ss = parseInt(tc_array[0])||0,
      tc_mm = parseInt(tc_array[1])||0,
      tc_hh = parseInt(tc_array[2])||0,
      tc_in_seconds = ( tc_hh * 3600 ) + ( tc_mm * 60 ) + tc_ss;
  return tc_in_seconds;
}```

File: dist/services/__test__/StructuralMetadataUtils.test.js (hunk 1 above)
Comment (Dananji) [reply dist/services/__test__/StructuralMetadataUtils.test.js]: No. I'll remove it :+1:

File: src/services/StructuralMetadataUtils.js [hunk 2]
```diff
@@ -78,25 +78,24 @@ export default class StructuralMetadataUtils {
 
     // Convert time to HH:mm:ss.ms format to use in validation logic
     let convertToSeconds = time => {
```
Comment (cjcolvar) [inline src/services/StructuralMetadataUtils.js]: It looks like the approach taken here is converging on what we have in our ruby code:
https://github.com/avalonmediasystem/avalon/blob/develop/app/helpers/application_helper.rb#L173-L182
Or here:
https://github.com/avalonmediasystem/avalon/blob/develop/app/controllers/application_controller.rb#L129-L147

**Comment by Dananji** [comment]
Thank you!! I guess I was not clear on the accepted time formatting in Avalon. I adopted the logic in the Ruby code. But still had to keep the regex tests because when there are missing begin/end times the SME gives `NaN` when this regex check is removed.

File: src/services/StructuralMetadataUtils.js [hunk 3]
```diff
@@ -68,35 +68,31 @@ export default class StructuralMetadataUtils {
    * @param {Float} duration - end time of the media file in Milliseconds
    */
   buildSMUI(allItems, duration) {
-    // Regex to match hh:mm:ss, mm:ss OR seconds(as a float)/minutes(as an int)
-    const regexHHMMSS = /^([0-9]*:[0-9]*:[0-9]*.[0-9]*)$/i;
+    // Regex to match hh:mm:ss.ms, mm:ss.ms OR minutes(as an int)
+    const regexHHMMSS = /^([0-9]*:[0-9]*(:|.)[0-9]*.[0-9]*)$/i;
```
Comment (bkeese) [inline src/services/StructuralMetadataUtils.js]: I think this one line change should be reverted. As is, it will match mm.ss.ms, which we don't want.

File: dist/services/StructuralMetadataUtils.js [hunk 4]
```diff
@@ -156,51 +156,43 @@ function () {
     value: function buildSMUI(allItems, duration) {
       var _this2 = this;
 
-      // Regex to match hh:mm:ss, mm:ss OR seconds(as a float)/minutes(as an int)
-      var regexHHMMSS = /^([0-9]*:[0-9]*:[0-9]*.[0-9]*)$/i;
-      var regexMMSS = /^([0-9]*:[0-9]*)$/i;
-      var regexSS = /^([0-9]*.[0-9]*)$/i; // Convert file duration to seconds
+      // Regex to match hh:mm:ss.ms, mm:ss.ms OR minutes(as an int)
+      var regexHHMMSS = /^([0-9]*:[0-9]*(:|.)[0-9]*.[0-9]*)$/i;
+      var regexMMSS = /^([0-9]*:[0-9]*)$/i; // Convert file duration to seconds
```
Comment (cjcolvar) [inline dist/services/StructuralMetadataUtils.js]: I think this is still not quite right since from my reading all times require a `:` so the `ss.sss` case isn't being supported.
I think in the end we need these patterns supported:
ss(.sss)
mm:ss(.sss)
hh:mm:ss(.sss)
so this might be more like this ruby regex:
`^(\d{0,2}:)?([012345]?\d:)?([012345]?\d)(\.\d{0,3})?$`
https://rubular.com/r/KJ9J4WNnOcQ6iD

File: src/services/StructuralMetadataUtils.js (hunk 3 above)
Comment (Dananji) [reply src/services/StructuralMetadataUtils.js]: It doesn't match `mm.ss.ms` format with this regex according to this, https://regex101.com/r/URiYI0/2/

Comment (bkeese) [reply src/services/StructuralMetadataUtils.js]: I should have been more specific. It will match hh:mm.ss.ms, which we don't want.

File: dist/services/StructuralMetadataUtils.js [hunk 5]
```diff
@@ -156,51 +156,40 @@ function () {
     value: function buildSMUI(allItems, duration) {
       var _this2 = this;
 
-      // Regex to match hh:mm:ss, mm:ss OR seconds(as a float)/minutes(as an int)
-      var regexHHMMSS = /^([0-9]*:[0-9]*:[0-9]*.[0-9]*)$/i;
-      var regexMMSS = /^([0-9]*:[0-9]*)$/i;
-      var regexSS = /^([0-9]*.[0-9]*)$/i; // Convert file duration to seconds
+      // Regex to match hh:mm:ss.ms / mm:ss.ms
+      var regexHHMMSS = /^([0-9]*:[0-9]*(:|.)[0-9]*.[0-9]*)$/i; // Convert file duration to seconds
```
Comment (bkeese) [inline dist/services/StructuralMetadataUtils.js]: We really don't want the `(:|.)`. It should just be `:`.

File: src/services/StructuralMetadataUtils.js (hunk 3 above)
Comment (bkeese) [reply src/services/StructuralMetadataUtils.js]: Same here: We really don't want the `(:|.)`. It should just be `:`.

File: dist/services/StructuralMetadataUtils.js (hunk 5 above)
Comment (Dananji) [reply dist/services/StructuralMetadataUtils.js]: If we replace `(:|.)` with just `:`, the tests for parsing `mm:ss` and `mm:ss.ms` fail.
Screenshot from 2020-01-07 17-08-08

Comment (cjcolvar) [reply dist/services/StructuralMetadataUtils.js]: I think that `/^([0-9]*:[0-9]*:[0-9]*.[0-9]*)$/` works well enough for HH:MM:SS.sss case but we need to handle the other possible scenarios as well (MM:SS.sss, HH:MM:SS, and MM:SS).  The SS and SS.sss cases are already handled by the `Number(time)` check. Here's the regex I propose to handle all of the cases at once: `/^(\d{1,2}:){0,2}(\d{1,2})(\.\d{0,3})?$/`
FWIW I think the `.` also needs to be escaped in the regex to avoid it matching any character.
It would probably be good to add tests for all of these cases and maybe some negative cases as well to ensure we don't match when we don't want to.

**Comment by cjcolvar** [comment]
@Dananji Sorry this is confusing with all of the comments!  Do you mind trying to talk through this tomorrow?  After talking with Brian, I think I'm in agreement with him that there might be a simpler solution that doesn't require a regex.

File: dist/services/StructuralMetadataUtils.js [hunk 6]
```diff
@@ -156,51 +152,39 @@ function () {
...
-              _seconds = _time$split$reverse2[0],
-              _minutes = _time$split$reverse2[1],
-              hours = _time$split$reverse2[2];
+          var _time$split$reverse2 = (0, _slicedToArray2["default"])(_time$split$reverse, 3);
 
+          seconds = _time$split$reverse2[0];
+          minutes = _time$split$reverse2[1];
+          hours = _time$split$reverse2[2];
           var hoursInS = hours ? parseInt(hours) * 3600 : 0;
+          var minutesInS = parseInt(minutes) * 60;
```
Comment (bkeese) [inline dist/services/StructuralMetadataUtils.js]: In case there are no minutes, I would handle these just like hours: have a default of 0.

File: src/services/StructuralMetadataUtils.js [hunk 7]
```diff
@@ -68,35 +66,29 @@ export default class StructuralMetadataUtils {
...
-        let [seconds, minutes, hours] = time.split(':').reverse();
+      let [hours, minutes, seconds] = [];
+      // Check ss/ss.ms values
+      if (Number(time)) {
+        let secondsNum = parseFloat(time);
+        return secondsNum;
+      } else {
+        [seconds, minutes, hours] = time.split(':').reverse();
         let hoursInS = hours ? parseInt(hours) * 3600 : 0;
         let minutesInS = parseInt(minutes) * 60;
```
Comment (bkeese) [inline src/services/StructuralMetadataUtils.js]: In case there are no minutes, I would handle these just like hours: have a default of 0.
